### PR TITLE
feat(computed): Computed accessing itself throw throws

### PR DESCRIPTION
**Pull request by HamadaBrest**
Fixes https://github.com/JsCommunity/reaclette/issues/19

### CODE REVIEW

File: src/factory.js [hunk 1]
```diff
@@ -131,6 +132,12 @@ module.exports = ({ Component, createElement, PropTypes }) => {
                       completeStateKeys,
                       stateAccessor
                     )
+                    Object.defineProperty(stateProxy, k, {
+                      enumerable: true,
+                      get () {
+                        throw new Error('cannot access computed from itself')
```
Comment (julien-f) [inline src/factory.js]: You should probably add the name of the computed in the message to help the user.

**Comment by julien-f** [comment]
Please add a test for this.

File: src/factory.js [hunk 2]
```diff
@@ -131,6 +132,12 @@ module.exports = ({ Component, createElement, PropTypes }) => {
                       completeStateKeys,
                       stateAccessor
                     )
+                    Object.defineProperty(stateProxy, k, {
+                      enumerable: true,
+                      get () {
+                        throw new Error(`computed: ${k}, cannot access computed from itself`)
```
Comment (julien-f) [inline src/factory.js]:
```
`computed "${k}" cannot depend on itself`
```

File: src/index.spec.js [hunk 3]
```diff
@@ -159,10 +159,12 @@ describe('provideState', () => {
 
   describe('computed', () => {
     const sum = jest.fn(({ foo }, { bar }) => foo + bar)
+    const circularComputed = jest.fn(({ circularComputed }, { bar }) => circularComputed + bar)
```
Comment (julien-f) [inline src/index.spec.js]: No need to use a prop, and you can leave the body empty, the state destructuring is enough :slightly_smiling_face:

File: src/index.spec.js [hunk 4]
```diff
@@ -197,6 +199,12 @@ describe('provideState', () => {
       expect(sum).toHaveBeenCalledTimes(3)
     })
 
+    it('throws when a computed calls its self', () => {
+      expect(() => {
+        return getInjectedState().circularComputed
+      }).toThrowError(new Error('computed: circularComputed, cannot access computed from itself'))
```
Comment (julien-f) [inline src/index.spec.js]: How will it work with #24?

Comment (HamadaBrest) [reply src/index.spec.js]: maybe we can type errors and do something like this
```js
try {
                    previousValue = c(stateProxy, propsProxy)
                  } catch (error) {
                    if (error instanceof ComputedError) {
                      throw error
                    }
                    noop()
                  }
```

Comment (julien-f) [reply src/index.spec.js]: I'm not very fond of this but I don't see better solutions.

File: src/reaclette-errors.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import makeError from 'make-error'
+
+export const ComputedError = makeError('ComputedError')
```
Comment (julien-f) [inline src/reaclette-errors.js]: Name the file `_CircularComputedError.js`

```js
import { BaseError } from 'make-error'

export default class CircularComputedError from BaseError {
  constructor(name) {
    super(`computed "${name}" cannot depend on itself`)
  }
}
```

File: src/factory.js [hunk 6]
```diff
@@ -131,13 +134,26 @@ module.exports = ({ Component, createElement, PropTypes }) => {
...
                   stateSpy.clear()
-                  previousValue = c(stateProxy, propsProxy)
+                  try {
+                    previousValue = c(stateProxy, propsProxy)
+                  } catch (error) {
+                    if (error instanceof CircularComputedError) {
+                      throw error
+                    }
+                    noop()
+                  }
```
Comment (julien-f) [inline src/factory.js]: This is not related to this PR.

File: src/factory.js [hunk 7]
```diff
@@ -131,13 +134,26 @@ module.exports = ({ Component, createElement, PropTypes }) => {
                       completeStateKeys,
                       stateAccessor
                     )
+                    Object.defineProperty(stateProxy, k, {
+                      enumerable: true,
```
Comment (julien-f) [inline src/factory.js]: Let's make it non-enumerable.

File: src/factory.js [hunk 8]
```diff
@@ -131,6 +134,11 @@ module.exports = ({ Component, createElement, PropTypes }) => {
                       completeStateKeys,
```
Comment (julien-f) [inline src/factory.js]: We could remove `k` from the spy because there is a special behavior on this, it will allow us to remove the `configurable` above.

Comment (HamadaBrest) [reply src/factory.js]: Do you mean like this : 
```js
                    const _completeStateKeys = { ...completeStateKeys }
                    delete _completeStateKeys[k];
                    [propsProxy, propsSpy] = makeSpy(propsKeys, propsAccessor);
                    [stateProxy, stateSpy] = makeSpy(
                      _completeStateKeys,
                      stateAccessor
                    )
                    // Removed code
                    // Object.defineProperty(stateProxy, k, {
                    //   enumerable: false,
                    //   get () {
                    //     throw new CircularComputedError(k)
                    //   },
                    // })
```

Comment (julien-f) [reply src/factory.js]: No.

What the type of `completeStateKeys`?

Comment (HamadaBrest) [reply src/factory.js]: my bad, it's an array, So normally when I remove the key from the `completeStateKeys` we will not need any extra code and we can put the as not configurable

Comment (HamadaBrest) [reply src/factory.js]:
```js
const computedIndex = completeStateKeys.indexOf(k)
                    if (computedIndex > -1) {
                      completeStateKeys.splice(computedIndex, 1)
                    }
                    [propsProxy, propsSpy] = makeSpy(propsKeys, propsAccessor);
                    [stateProxy, stateSpy] = makeSpy(
                      _completeStateKeys,
                      stateAccessor
                    )
                    // Removed code
                    // Object.defineProperty(stateProxy, k, {
                    //   enumerable: false,
                    //   get () {
                    //     throw new CircularComputedError(k)
                    //   },
                    // })
```

Comment (julien-f) [reply src/factory.js]: Nope, this will break everything.

You should be able to figure out why.

Comment (HamadaBrest) [reply src/factory.js]: My bad, I meant to splice the copied array and not main one

```js
                   const _completeStateKeys = completeStateKeys
                    const computedIndex = completeStateKeys.indexOf(k)
                    if (computedIndex > -1) {
                      _completeStateKeys.splice(computedIndex, 1)
                    }
                    [propsProxy, propsSpy] = makeSpy(propsKeys, propsAccessor);
                    [stateProxy, stateSpy] = makeSpy(
                      _completeStateKeys,
                      stateAccessor
                    )
```

Comment (julien-f) [reply src/factory.js]: Why so complicated instead of simply using [`Array#filter`](https://devdocs.io/javascript/global_objects/array/filter)?

Comment (HamadaBrest) [reply src/factory.js]: I don't get the idea for this comment https://github.com/JsCommunity/reaclette/pull/23#discussion_r261967165
By removing K from the spy, we remove tracibility change for the computed named `K` and it becomes undefined.
Why do you want to do that ?

Comment (julien-f) [reply src/factory.js]: We don't need to spy the property `k` because it will not be accessed (should not) from the computed of the same name.

That's exactly what you are working on in this PR.

Comment (HamadaBrest) [reply src/factory.js]: I understand, but making `k` not accessible inside its computed makes things unclear for the developer who calls a computed in it self, don't you think so ?

File: src/factory.js [hunk 9]
```diff
@@ -126,11 +129,19 @@ module.exports = ({ Component, createElement, PropTypes }) => {
               stateDescriptors[k] = {
                 get: () => {
                   if (propsProxy === undefined) {
+                    const _completeStateKeys = completeStateKeys.filter(key => key !== k);
                     [propsProxy, propsSpy] = makeSpy(propsKeys, propsAccessor);
                     [stateProxy, stateSpy] = makeSpy(
-                      completeStateKeys,
+                      _completeStateKeys,
```
Comment (julien-f) [inline src/factory.js]: Inline this.

File: src/factory.js [hunk 10]
```diff
@@ -11,6 +13,7 @@ const makeSpy = (keys, accessor) => {
   const spy = new Map()
   keys.forEach(k => {
     descriptors[k] = {
+      configurable: false,
```
Comment (julien-f) [inline src/factory.js]: No longer necessary, because we don't re-define this property anymore.

File: src/factory.js [hunk 11]
```diff
@@ -126,11 +129,19 @@ module.exports = ({ Component, createElement, PropTypes }) => {
...
+                    const _completeStateKeys = completeStateKeys.filter(key => key !== k);
                     [propsProxy, propsSpy] = makeSpy(propsKeys, propsAccessor);
                     [stateProxy, stateSpy] = makeSpy(
-                      completeStateKeys,
+                      _completeStateKeys,
                       stateAccessor
                     )
+
+                    Object.defineProperty(stateProxy, k, {
+                      enumerable: false,
```
Comment (julien-f) [inline src/factory.js]: `false` is the default value, this line is unnecessary.
